perf/rdm_bw.c

```diff
@@ -0,0 +1,471 @@
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <assert.h>
+#include <sys/time.h>
+#include <string.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_rma.h>
+
+#include "ft_utils.h"
+#include "../include/shared.h"
+
+#define MAX_REQ_NUM 1000
+
+#define MAX_ALIGNMENT 65536
+#define MAX_MSG_SIZE (1<<22)
+#define MYBUFSIZE (MAX_MSG_SIZE + MAX_ALIGNMENT)
+
+#define HEADER "# Libfabric Bandwidth Test \n"
+#ifndef FIELD_WIDTH
+#   define FIELD_WIDTH 20
+#endif
+#ifndef FLOAT_PRECISION
+#   define FLOAT_PRECISION 2
+#endif
+
+int loop = 100;
+int window_size = 64;
+int skip = 10;
+
+int loop_large = 20;
+int window_size_large = 64;
+int skip_large = 2;
+
+int large_message_size = 8192;
+
+char s_buf_original[MYBUFSIZE];
+char r_buf_original[MYBUFSIZE];
+
+static int rx_depth = 512;
+
+struct fi_info *fi, *hints;
+struct fid_fabric *fab;
+struct fid_domain *dom;
+struct fid_ep *ep;
+struct fid_av *av;
+struct fid_cq *rcq, *scq;
+struct fid_mr *r_mr, *l_mr;
+struct fi_context fi_ctx_send;
+struct fi_context fi_ctx_recv;
+struct fi_context fi_ctx_av;
+
+void *addrs;
+fi_addr_t *fi_addrs;
+
+typedef struct buf_desc {
+	uint64_t addr;
+	uint64_t key;
+} buf_desc_t;
+
+buf_desc_t *rbuf_descs;
+
+int myid, numprocs;
+
+void print_usage(void)
+{
+	fprintf(stderr, "\nOptions:\n");
+	fprintf(stderr, "  -f <provider>\tspecific provider name eg IP, verbs\n");
+	fprintf(stderr, "  -h\t\tdisplay this help output\n");
+	return;
+}
+
+static double get_time(void)
+{
+	struct timeval tv;
+	double td;
+	gettimeofday(&tv, NULL);
+	td = tv.tv_sec + (tv.tv_usec * 0.000001);
+	return td;
+}
```

Comment:
This functionality should be in common code, though I would rather see it return a u64 as micro-seconds or nano-seconds.